README
README file for GraphViz2.

See also: CHANGES and Changelog.ini.

Warning: WinZip 8.1 and 9.0 both contain an 'accidental' bug which stops
them recognizing POSIX-style directory structures in valid tar files.
You are better off using a reliable tool such as InfoZip:
ftp://ftp.info-zip.org/pub/infozip/

1 Installing from a Unix-like distro
------------------------------------
shell>gunzip GraphViz2-1.00.tgz
shell>tar mxvf GraphViz2-1.00.tar

On Unix-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>./Build
shell>./Build test
shell>./Build install

On MS Windows-like systems, assuming you have installed Module::Build V 0.25+:

shell>perl Build.PL
shell>perl Build
shell>perl Build test
shell>perl Build install

Alternately, without Module::Build, you do this:

Note: 'make' on MS Windows-like systems may be called 'nmake' or 'dmake'.

shell>perl Makefile.PL
shell>make
shell>make test
shell>su              (for Unix-like systems)
shell>make install
shell>exit            (for Unix-like systems)

On all systems:

Run GraphViz2.pm through you favourite pod2html translator.

2 Installing from an ActiveState distro
---------------------------------------
shell>unzip GraphViz2-1.00.zip
shell>ppm install --location=. GraphViz2
shell>del GraphViz2-1.00.ppd
shell>del PPM-GraphViz2-1.00.tar.gz


CHANGES
Revision history for Perl extension GraphViz2.

1.12  Wed Dec 14 11:22:00 2011
	  - Add 5 scripts, scripts/macro.*.pl, to demonstrate using a Perl sub as a macro to generate both cluster and non-cluster sub-graphs.
	  - Update the TODO in GraphViz2.pm, since it erroneously gave the impression the current code did not handle compound clusters.
	  - Re-generate the demo. See: http://savage.net.au/Perl-modules/html/graphviz2/.
	  - Adopt Flavio Poletti's suggestion of trying to pipe to dot, in Build.PL/Makefile.PL, rather than using File::Which,
	  	to see if dot (Graphviz) is installed. This (hopefully) solves the problem of using File::Which on systems where it is
		not installed, before Build.PL/Makefile.PL has a chance to tell the user that File::Which is required. See: RT#73077.
	  - No code changes in *.pm files.

1.11  Tue Nov  8 10:21:00 2011
	  - Fix bug where double quotes in HTML labels were being escaped and should not have been.
	  	Many thanx to Fitz Elliott for the report and initial patch. See https://rt.cpan.org/Ticket/Display.html?id=72259.
	  - Patched scripts/html.labels.pl as per Fitz's suggested test code.
	  - Rename t/lib/Parent.pm to t/lib/Adult.pm to avoid a reported problem with "use parent 'Parent';" under Windows.
	  	This affect various files. See https://rt.cpan.org/Public/Bug/Display.html?id=69076.

1.10  Thu Sep  1 10:45:00 2011
	  - Fix bug in GraphViz in handling of nodes whose shape is none or plaintext, and whose label is HTML.
	  - Fix bug in scripts/generate.png.sh to add png parameter to 'perl -Ilib scripts/generate.demo.pl png'.
	  - HTML labels work without any code changes, so references to them not working have been removed from the docs.
	  - GraphViz2 is called by the now-released Graph::Easy::Marpa, so remove that item from the TODO list.
	  - Fix doc referring to scripts/generate.index.pl. It now refers to scripts/generate.demo.pl.
	  - Add DOCTYPE and Content-type to html/graphviz.index.tx. Output by scripts/generate.demo.pl, html/index.html, now handles UTF-8 properly.
	  - Add alt to the <img ...> tags output by scripts/generate.demo.pl.
	  - Add scripts/html.labels.pl. This code includes a demo of using ports.
	  - Reorder methods in GraphViz2::Utils to be in alphabetical order.

1.09  Thu Jul 21 14:13:00 2011
	  - Patch parameter validation to allow for output image types such as png:gd etc.
	  	Only the prefix before the first ':' is validated.

1.08  Thu Jul 21 11:53:00 2011
	  - Change the behaviour of sub log() in GraphViz.pm. Now if called with $level eq 'error', it dies with $message.
	  - Change references to the color darkblue to blue, so it's compatible with both the X11 and SVG color schemes.
	  - Likewise change chartreuse to lawngreen.
	  - Rename scripts/generate.index.pl to scripts/generate.demo.pl, as a more meaningful name.
	  - Patch scripts/generate.demo.pl to accept the output image type as a command line parameter.
	  - Use this new feature to generate png files, and upload them to the demo site
	  	http://savage.net.au/Perl-modules/html/graphviz2/ even though png is often uglier than svg.
		I did this because Iceweasel (Firefox) V 3.5.16 was not displaying svgs within the index file,
		even though they are fine when displayed directly.
	  - Add scripts/generate.png.sh.
	  - Note: scripts/jointed.edges.pl must call dot with -Tpng:gd and not -Tpng, for unknown reasons.
	  - Make both scripts/generate.png.sh and scripts/generate.svg.sh redirect their log files to /tmp,
	  	so that we don't have to ship the logs, and also so they don't keep changing and hence need checking in.
	  - Remove ./dbi.schema.log from the distro.

1.07  Mon Jul  4 15:46:00 2011
	  - Use Date::Format to add a date stamp at the end of html/index.html, as output by generate.index.pl.

1.06  Tue Jun 28 11:10:00 2011
	  - Change usage of File::Temp -> newdir to fix problems testing on BSD-based systems.
	  - Add scripts/jointed.edges.pl.
	  	This demo - in Graph::Easy syntax - ships with Graph::Easy::Marpa.
	  - Re-write generate.index.pl to put all demo data (where available) and images on 1 page.
	  - Upload demo to (new location) http://savage.net.au/Perl-modules/html/graphviz2/index.html.

1.05  Fri Jun 24 12:40:00 2011
	  - Implement GraphViz2::Parse::Marpa, along with scripts/parse.marpa.pl and t/sample.marpa.1.dat.
	  	The output is html/parse.marpa.svg.
	  - Implement GraphViz2::Parse::STT, along with scripts/parse.stt.pl and t/sample.stt.1.dat.
	  	The output is html/parse.stt.svg.
	  - Add use File::Spec to t/test.t

1.04  Wed Jun 22  9:36:00 2011
	  - Reduce required version of File::Basename to 2.77, which came with Perl 5.10.1.
	  - Stop trying to write to t/html/, and use File::Temp for a directory instead.
	  	That way, it doesn't matter who owns t/html/, nor whether or not it's writable.

1.03  Sun Jun 19 16:27:00 2011
	  - Tweak File::Temp -> new to be File::Temp ->new(EXLOCK => 0) for BSD-based systems.

1.02  Fri Jun 17  8:36:00 2011
	  - Add the pre-requisite Log::Handler to Build.PL and Makefile.PL.
	  - Release HTML::Entities::Interpolate V 1.04 and Set::Array V 0.23 to CPAN.
	  - Add README file.
	  - Clean up TODO list.

1.01  Wed Jun 15 15:00:00 2011
	  - Quote cluster/subgraph names so they may contain weird characters.
	  - Add method dependency(data => $depend) to GraphViz2.pm, which accepts an object of type
	  	Algorithm::Dependency. See scripts/dependency.pl and html/dependency.svg.
	  - Add GraphViz2::Parse::ISA, and scripts/parse.isa.pl, and the t/lib/Parent hierarchy.

1.00  Wed Jun 15 14:26:00 2011
	  - This is a re-write of GraphViz. The method parameter lists are incompatible.
	  	Sorry, but it now supports all options and attributes in Graphviz V 2.23.6.
	  - Rewrite GraphViz, GraphViz::Data::Grapher, GraphViz::Parse::RecDescent, GraphViz::Parse::Yacc and GraphViz::Parse::Yapp.
	  	The core code of *::RecDescent, *::Yacc and *::Yapp has been copied from GraphViz, with tiny changes.

lib/GraphViz2/DBI.pm
package GraphViz2::DBI;

use strict;
use warnings;

use GraphViz2;

use Hash::FieldHash ':all';

fieldhash my %catalog    => 'catalog';
fieldhash my %dbh        => 'dbh';
fieldhash my %graph      => 'graph';
fieldhash my %schema     => 'schema';
fieldhash my %table      => 'table';
fieldhash my %table_info => 'table_info';
fieldhash my %type       => 'type';

our $VERSION = '1.12';

# -----------------------------------------------

sub create
{
	my($self, %arg) = @_;
	my($name) = $arg{name} || '';

	$self -> get_table_info;

	my($table_info) = $self -> table_info;

	my($port, %port);

	for my $table_name (sort keys %$table_info)
	{
		# Port 1 is the table name.

		$port              = 1;
		$port{$table_name} = {};

		for my $column_name (@{$$table_info{$table_name}{column_names} })
		{
			$port{$table_name}{$column_name} = $port++;
		}
	}

	for my $table_name (sort keys %$table_info)
	{
		$self -> graph -> add_node(name => $table_name, label => [$table_name, join('\l', map{"$port{$table_name}{$_}: $_"} sort keys %{$port{$table_name} }) . '\l']);
	}

	for my $table_name (sort keys %$table_info)
	{
		for my $item (sort @{$$table_info{$table_name}{foreign_keys} })
		{
			$self -> graph -> add_edge(from => "$table_name:port2", to => "$$item[1]:port2");
		}
	}

	if ($name)
	{
		$self -> graph -> add_node(name => $name, shape => 'doubleoctagon');

		for my $table_name (sort keys %$table_info)
		{
			$self -> graph -> add_edge(from => $name, to => $table_name);
		}
	}

	return $self;

} # End of create.

# -----------------------------------------------

sub get_table_info
{
	my($self)               = @_;
	my($dbh)                = $self -> dbh;
	$$dbh{FetchHashKeyName} = 'NAME_lc';
	my($vendor)             = uc $dbh -> get_info(17); # SQL_DBMS_NAME.
	my($sth)                = $dbh -> table_info($self -> catalog, $self -> schema, $self -> table, $self -> type);
	my($table_info)         = $sth -> fetchall_arrayref({});

	my($column_sth, @column_name);
	my($table_name, %table_data);

	for my $item (@$table_info)
	{
		$table_name = $$item{'table_name'};

		next if ( ($vendor eq 'ORACLE')     && ($table_name =~ /^bin\$.+\$./) );
		next if ( ($vendor eq 'POSTGRESQL') && ($table_name =~ /^(?:pg_|sql_)/) );
		next if ( ($vendor eq 'SQLITE')     && ($table_name eq 'sqlite_sequence') );

		my($column_sth) = $dbh -> prepare("select * from $table_name where 1 = 0");

		$column_sth -> execute;

		@column_name             = @{$$column_sth{NAME} };
		$table_data{$table_name} =
			{
				column_names => [sort @column_name],
				foreign_keys => [],
			};

		$column_sth -> finish;
	}

	my($column_data);
	my(@foreign_info);

	for my $table_name (sort keys %table_data)
	{
		@foreign_info = ();
		$sth          = $dbh -> foreign_key_info(undef, undef, undef, $self -> catalog, $self -> schema, $table_name);

		if (defined $sth)
		{
			$table_info = $sth -> fetchall_arrayref({});

			for $column_data (@$table_info)
			{
				push @foreign_info, [$$column_data{'fk_column_name'}, $$column_data{'uk_table_name'}, $$column_data{'uk_column_name'}];
			}

			$table_data{$table_name}{foreign_keys} = [sort @foreign_info];
		}
	}

	$self -> table_info(\%table_data);

} # End of get_table_info.

# -----------------------------------------------

sub _init
{
	my($self, $arg) = @_;
	$$arg{catalog}  ||= undef; # Caller can set.
	$$arg{dbh}      = $$arg{dbh} || die 'Error: You must supply a dbh';
	$$arg{graph}    ||= GraphViz2 -> new
		(
		 edge   => {color => 'grey'},
		 global => {directed => 1},
		 graph  => {rankdir => 'TB'},
		 logger => '',
		 node   => {color => 'blue', shape => 'oval'},
		);
	$$arg{schema}     ||= undef;   # Caller can set.
	$$arg{table}      ||= '%';     # Caller can set.
	$$arg{table_info} = {};
	$$arg{type}       ||= 'TABLE'; # Caller can set.
	$self             = from_hash($self, $arg);

	return $self;

} # End of _init.

# -----------------------------------------------

sub new
{
	my($class, %arg) = @_;
	my($self)        = bless {}, $class;
	$self            = $self -> _init(\%arg);

	return $self;

}	# End of new.

# -----------------------------------------------

1;

=pod

=head1 NAME

L<GraphViz2::DBI> - Visualize a database schema as a graph

=head1 Synopsis

	#!/usr/bin/env perl
	
	use strict;
	use warnings;
	
	use DBI;
	
	use GraphViz2;
	use GraphViz2::DBI;
	
	use Log::Handler;
	
	# ---------------
	
	exit 0 if (! $ENV{DBI_DSN});
	
	my($logger) = Log::Handler -> new;
	
	$logger -> add
		(
		 screen =>
		 {
			 maxlevel       => 'debug',
			 message_layout => '%m',
			 minlevel       => 'error',
		 }
		);
	
	my($graph) = GraphViz2 -> new
		(
		 edge   => {color => 'grey'},
		 global => {directed => 1},
		 graph  => {rankdir => 'TB'},
		 logger => $logger,
		 node   => {color => 'blue', shape => 'oval'},
		);
	my($dbh) = DBI -> connect($ENV{DBI_DSN}, $ENV{DBI_USER}, $ENV{DBI_PASS});
	my($g)   = GraphViz2::DBI -> new(dbh => $dbh, graph => $graph);
	
	$g -> create(name => 'App-Office-CMS');
	
	my($format)      = shift || 'svg';
	my($output_file) = shift || File::Spec -> catfile('html', "dbi.schema.$format");
	
	$graph -> run(format => $format, output_file => $output_file);

See scripts/dbi.schema.pl (L<GraphViz2/Scripts Shipped with this Module>).

=head1 Description

Takes a database handle, and graphs the schema.

You can write the result in any format supported by L<Graphviz|http://www.graphviz.org/>.

Here is the list of L<output formats|http://www.graphviz.org/content/output-formats>.

=head1 Distributions

This module is available as a Unix-style distro (*.tgz).

See L<http://savage.net.au/Perl-modules/html/installing-a-module.html>
for help on unpacking and installing distros.

=head1 Installation

Install L<GraphViz2> as you would for any C<Perl> module:

Run:

	cpanm GraphViz2

or run:

	sudo cpan GraphViz2

or unpack the distro, and then either:

	perl Build.PL
	./Build
	./Build test
	sudo ./Build install

or:

	perl Makefile.PL
	make (or dmake or nmake)
	make test
	make install

=head1 Constructor and Initialization

=head2 Calling new()

C<new()> is called as C<< my($obj) = GraphViz2::DBI -> new(k1 => v1, k2 => v2, ...) >>.

It returns a new object of type C<GraphViz2::DBI>.

Key-value pairs accepted in the parameter list:

=over 4

=item o dbh => $dbh

This options specifies the database handle to use.

This key is mandatory.

=item o graph => $graphviz_object

This option specifies the GraphViz2 object to use. This allows you to configure it as desired.

The default is GraphViz2 -> new. The default attributes are the same as in the synopsis, above,
except for the graph label of course.

This key is optional.

=back

=head1 Methods

=head2 create(name => $name)

Creates the graph, which is accessible via the graph() method, or via the graph object you passed to new().

Returns $self to allow method chaining.

$name is the string which will be placed in the root node of the tree. It may be omitted, in which case the root node is omitted.

=head2 graph()

Returns the graph object, either the one supplied to new() or the one created during the call to new().

=head1 FAQ

See L<GraphViz2/FAQ> and L<GraphViz2/Scripts Shipped with this Module>.

=head1 Thanks

Many thanks are due to the people who chose to make L<Graphviz|http://www.graphviz.org/> Open Source.

And thanks to L<Leon Brocard|http://search.cpan.org/~lbrocard/>, who wrote L<GraphViz>, and kindly gave me co-maint of the module.

=head1 Version Numbers

Version numbers < 1.00 represent development versions. From 1.00 up, they are production versions.

=head1 Machine-Readable Change Log

The file CHANGES was converted into Changelog.ini by L<Module::Metadata::Changes>.

=head1 Support

Email the author, or log a bug on RT:

L<https://rt.cpan.org/Public/Dist/Display.html?Name=GraphViz2>.

=head1 Author

L<GraphViz2> was written by Ron Savage I<E<lt>ron@savage.net.auE<gt>> in 2011.

Home page: L<http://savage.net.au/index.html>.

=head1 Copyright

Australian copyright (c) 2011, Ron Savage.

	All Programs of mine are 'OSI Certified Open Source Software';
	you can redistribute them and/or modify them under the terms of
	The Artistic License, a copy of which is available at:
	http://www.opensource.org/licenses/index.html

=cut
